File: CSharp_projects/AtmSystemModel/src/project5/Application/project5.Application/Accounts/AccountService.cs
using Lab5.Application.Abstractions.Repositories;
using Lab5.Application.Contracts;
using Lab5.Application.Contracts.Accounts;
using Lab5.Application.Models.Accounts;
using Lab5.Application.Models.Transactions;
using System.Collections.ObjectModel;

namespace Lab5.Application.Accounts;

public class AccountService : IAccountService
{
    private readonly IAccountRepository _accountRepository;

    public Account? Account { get; set; }

    public AccountService(IAccountRepository accountRepository)
    {
        _accountRepository = accountRepository;
        Account = null;
    }

    public LoginResult Login(long number, string password)
    {
        Account? account = _accountRepository.FindAccountByNumber(number);
        if (account?.Password == password)
        {
            Account = account;

            return new LoginResult.Success();
        }

        return new LoginResult.NotFound();
    }

    public void Refill(int amount)
    {
        if (Account != null)
        {
            Account.Balance += amount;
            var list = Account.Transactions.ToList();
            list.Add(new Transaction(TransactionType.Refill, amount));

            Account.Transactions = list;
            _accountRepository.UpdateAccount(Account);
        }
    }

    public double CheckBalance()
    {
        if (Account != null)
        {
            var list = Account.Transactions.ToList();
            list.Add(new Transaction(TransactionType.Check, 0));
            return Account.Balance;
        }

        return 0;
    }

    public void Withdraw(int amount)
    {
        if (Account != null && Account.Balance > amount)
        {
            Account.Balance -= amount;
            var list = Account.Transactions.ToList();
            list.Add(new Transaction(TransactionType.Withdraw, amount));

            Account.Transactions = list;
            _accountRepository.UpdateAccount(Account);
        }
    }

    public IEnumerable<Transaction> GetTransactionHistory()
    {
        if (Account != null)
        {
            return (ReadOnlyCollection<Transaction>)Account.Transactions;
        }

        return new ReadOnlyCollection<Transaction>(new List<Transaction>());
    }
}

File: Java_projects/project1/src/main/java/com/project1/services/AccountService.java
package com.project1.services;

import com.project1.exceptions.NotEnoughMoneyException;
import com.project1.exceptions.NotFoundAccountException;
import com.project1.models.Account;
import com.project1.models.transactions.Transaction;
import com.project1.models.transactions.TransactionType;
import com.project1.repositories.IAccountRepository;

import java.util.ArrayList;
import java.util.List;

/**
 * Реализация сервиса для работы с аккаунтами.
 */
public class AccountService  {
    private final IAccountRepository accountRepository;
    private Account account;

    /**
     * Конструктор сервиса.
     *
     * @param accountRepository репозиторий аккаунтов
     */
    public AccountService(IAccountRepository accountRepository) {
        this.accountRepository = accountRepository;
        this.account = null;
    }

    /**
     * Выполняет вход в аккаунт.
     *
     * @param number   номер аккаунта
     * @param password пароль аккаунта
     * @return найденный аккаунт
     */
    public Account Login(long number, String password) throws NotFoundAccountException {
        Account foundAccount = accountRepository.findAccountByNumber(number);
        if (foundAccount != null && foundAccount.getPassword().equals(password)) {
            this.account = foundAccount;
            return foundAccount;
        }
        throw new NotFoundAccountException("Неверный номер счета или пароль.");
    }

    /**
     * Пополняет баланс авторизованного аккаунта.
     *
     * @param amount сумма пополнения
     */
    public void refill(int amount) throws NotFoundAccountException {
        if (account == null) {
            throw new NotFoundAccountException("Аккаунт не авторизован.");
        }
        account.setBalance(account.getBalance() + amount);
        List<Transaction> transactions = new ArrayList<>(account.getTransactions());
        transactions.add(new Transaction(TransactionType.REFILL, amount));
        account.setTransactions(transactions);
        accountRepository.updateAccount(account);
    }

    /**
     * Возвращает баланс авторизованного аккаунта.
     *
     * @return текущий баланс
     */
    public double checkBalance() throws NotFoundAccountException {
        if (account == null) {
            throw new NotFoundAccountException("Аккаунт не авторизован.");
        }
        List<Transaction> transactions = new ArrayList<>(account.getTransactions());
        transactions.add(new Transaction(TransactionType.CHECK, 0));
        return account.getBalance();
    }

    /**
     * Снимает указанную сумму с баланса авторизованного аккаунта.
     *
     * @param amount сумма для снятия
     */
    public void withdraw(int amount) throws NotFoundAccountException, NotEnoughMoneyException {
        if (account == null) {
            throw new NotFoundAccountException("Аккаунт не авторизован.");
        }
        if (account.getBalance() <= amount) {
            throw new NotEnoughMoneyException("Недостаточно средств.");
        }
        account.setBalance(account.getBalance() - amount);
        List<Transaction> transactions = new ArrayList<>(account.getTransactions());
        transactions.add(new Transaction(TransactionType.WITHDRAW, amount));
        account.setTransactions(transactions);
        accountRepository.updateAccount(account);
    }

    /**
     * Возвращает историю транзакций авторизованного аккаунта.
     *
     * @return список транзакций
     */
    public List<Transaction> getTransactionHistory() throws NotFoundAccountException {
        if (account == null) {
            throw new NotFoundAccountException("Аккаунт не авторизован.");
        }
        return new ArrayList<>(account.getTransactions());
    }
}
